File: ZeroTier One/NetworkMonitor.m
//
//  NetworkMonitor.m
//  ZeroTier One
//
//  Created by Grant Limberg on 8/7/16.
//  Copyright © 2016 ZeroTier, Inc. All rights reserved.
//

#import "NetworkMonitor.h"
#import "Network.h"
#import "ServiceCom.h"
#import "NodeStatus.h"

NSString * const NetworkUpdateKey = @"com.zerotier.one.network-list";
NSString * const StatusUpdateKey = @"com.zerotier.one.status";

@interface NetworkMonitor (private)

- (NSString*)dataFile;
- (void)internal_updateNetworkInfo;
- (NSInteger)findNetworkWithID:(UInt64)networkId;
- (NSInteger)findSavedNetworkWithID:(UInt64)networkId;
- (void)saveNetworks;

@end

@implementation NetworkMonitor

- (id)init
{
    self = [super init];
    if(self)
    {
        _savedNetworks = [NSMutableArray<Network*> array];
        _receivedNetworks = [NSArray<Network*> array];
        _allNetworks = [NSMutableArray<Network*> array];
        _timer = nil;
    }

    return self;
}

- (void)dealloc
{
    [_timer invalidate];
}

- (void)start
{
    NSLog(@"ZeroTier monitor started");
    _timer = [NSTimer scheduledTimerWithTimeInterval:1.0f
                                              target:self
                                            selector:@selector(updateNetworkInfo)
                                            userInfo:nil
                                             repeats:YES];
}

- (void)stop
{
    NSLog(@"ZeroTier monitor stopped");
    [_timer invalidate];
    _timer = nil;
}

- (void)updateNetworkInfo
{
    NSString *filePath = [self dataFile];

    if([[NSFileManager defaultManager] fileExistsAtPath:filePath]) {
        NSArray<Network*> *networks = [NSKeyedUnarchiver unarchiveObjectWithFile:filePath];

        if(networks != nil) {
            _savedNetworks = [networks mutableCopy];
        }
    }

    NSError *error = nil;

    [[ServiceCom sharedInstance] getNetworklist:^(NSArray<Network *> *networkList) {
        _receivedNetworks = networkList;

        [[NSOperationQueue mainQueue] addOperationWithBlock:^{
            [self internal_updateNetworkInfo];
        } ];
    } error:&error];

    if(error) {
        // TODO: Display error message

        [self stop];
    }

    [[ServiceCom sharedInstance] getNodeStatus:^(NodeStatus *status) {
        NSDictionary *userInfo = [NSDictionary dictionaryWithObject:status forKey:@"status"];

        [[NSOperationQueue mainQueue] addOperationWithBlock:^{
            [[NSNotificationCenter defaultCenter] postNotificationName:StatusUpdateKey
                                                                object:nil
                                                              userInfo:userInfo];
        }];
    } error:&error];

    if (error) {
        // TODO: Display error message

        [self stop];
    }
}

- (void)deleteSavedNetwork:(NSString*)networkId
{
    UInt64 nwid = 0;
    NSScanner *scanner = [NSScanner scannerWithString:networkId];
    [scanner scanHexLongLong:&nwid];

    NSInteger index = [self findNetworkWithID:nwid];

    if(index != NSNotFound) {
        [_allNetworks removeObjectAtIndex:index];
    }

    index = [self findSavedNetworkWithID:nwid];

    if(index != NSNotFound) {
        [_savedNetworks removeObjectAtIndex:index];
    }

    [self saveNetworks];
}

@end

@implementation NetworkMonitor (private)
- (NSString*)dataFile
{
    NSURL *appSupport = [[[NSFileManager defaultManager] URLsForDirectory:NSApplicationSupportDirectory
                                                               inDomains:NSUserDomainMask] objectAtIndex:0];

    appSupport = [[[appSupport URLByAppendingPathComponent:@"ZeroTier"] URLByAppendingPathComponent:@"One"] URLByAppendingPathComponent:@"networkinfo.dat"];
    return appSupport.path;
}

- (void)internal_updateNetworkInfo
{
    NSMutableArray<Network*> *networks = [_savedNetworks mutableCopy];

    for(Network *nw in _receivedNetworks) {
        NSInteger index = [self findSavedNetworkWithID:nw.nwid];

        if(index != NSNotFound) {
            [networks setObject:nw atIndexedSubscript:index];
        }
        else {
            [networks addObject:nw];
        }
    }

    [networks sortUsingComparator:^NSComparisonResult(Network *obj1, Network *obj2) {
        if(obj1.nwid > obj2.nwid) {
            return true;
        }
        return false;
    }];

    @synchronized(_allNetworks) {
        _allNetworks = networks;
    }

    [self saveNetworks];

    NSDictionary *userInfo = [NSDictionary dictionaryWithObject:networks forKey:@"networks"];

    [[NSNotificationCenter defaultCenter] postNotificationName:NetworkUpdateKey
                                                        object:nil
                                                      userInfo:userInfo];
}

- (NSInteger)findNetworkWithID:(UInt64)networkId
{
    for(int i = 0; i < [_allNetworks count]; ++i) {
        Network *nw = [_allNetworks objectAtIndex:i];

        if(nw.nwid == networkId) {
            return i;
        }
    }

    return NSNotFound;
}


- (NSInteger)findSavedNetworkWithID:(UInt64)networkId
{
    for(int i = 0; i < [_savedNetworks count]; ++i) {
        Network *nw = [_savedNetworks objectAtIndex:i];

        if(nw.nwid == networkId) {
            return i;
        }
    }

    return NSNotFound;
}

- (void)saveNetworks
{
    NSString *filePath = [self dataFile];

    @synchronized(_allNetworks) {
        [NSKeyedArchiver archiveRootObject:_allNetworks toFile:filePath];
    }
}

@end

File: ZeroTier One/NetworkMonitor.swift
//
//  NetworkMonitor.swift
//  ZeroTier One
//
//  Created by Grant Limberg on 6/16/16.
//  Copyright © 2016 ZeroTier, Inc. All rights reserved.
//

import Cocoa

let networkUpdateKey = "com.zerotier.one.network-list"
let statusUpdateKey = "com.zerotier.one.status"

class NetworkMonitor: NSObject {

    var timer: NSTimer? = nil

    var savedNetworks: [Network] = [Network]()
    var receivedNetworks: [Network] = [Network]()
    var allNetworks: [Network] = [Network]()

    override init() {
        super.init()
    }

    deinit {
        timer?.invalidate()
    }

    func start() {
        NSLog("ZeroTier monitor started")
        timer = NSTimer.scheduledTimerWithTimeInterval(1.0,
                                                       target: self,
                                                       selector: #selector(updateNetworkInfo),
                                                       userInfo: nil,
                                                       repeats: true)
    }


    func stop() {
        NSLog("ZeroTier monitor stopped")
        timer?.invalidate()
        timer = nil
    }

    private func  dataFile() -> String {
        var appSupport = NSFileManager.defaultManager().URLsForDirectory(.ApplicationSupportDirectory, inDomains: .UserDomainMask)[0]
        appSupport = appSupport.URLByAppendingPathComponent("ZeroTier").URLByAppendingPathComponent("One").URLByAppendingPathComponent("networks.dat")
        return appSupport.path!
    }

    func updateNetworkInfo() {
        let filePath = dataFile()

        if NSFileManager.defaultManager().fileExistsAtPath(filePath) {
            let networks = NSKeyedUnarchiver.unarchiveObjectWithFile(filePath) as! [Network]

            self.savedNetworks.removeAll()

            for n in networks {
                self.savedNetworks.append(n)
            }
        }

        ServiceCom.getNetworkList() { (networkList) -> Void in
            self.receivedNetworks = networkList

            NSOperationQueue.mainQueue().addOperationWithBlock() { () -> Void in
                self.internal_updateNetworkInfo()
            }
        }

        ServiceCom.getNodeStatus() { nodeStatus -> Void in
            NSOperationQueue.mainQueue().addOperationWithBlock() { () -> Void in
                let nc = NSNotificationCenter.defaultCenter()

                nc.postNotificationName(statusUpdateKey, object: nil, userInfo: ["status": nodeStatus])
            }
        }
    }

    func deleteSavedNetwork(nwid: String) {
        if let nwid = UInt64(nwid, radix: 16) {
            let index = findNetworkWithID(nwid)

            if index != NSNotFound {
                allNetworks.removeAtIndex(index)
            }

            let index2 = findSavedNetworkWithID(nwid)

            if index2 != NSNotFound {
                savedNetworks.removeAtIndex(index2)
            }
        }

        saveNetworks()
    }

    // Only to be called by updateNetworkInfo()
    private func internal_updateNetworkInfo() {
        var networks = self.savedNetworks

        for nw in receivedNetworks {
            let index = findSavedNetworkWithID(nw.nwid)

            if index != NSNotFound {
                networks[index] = nw
            }
            else {
                networks.append(nw)
            }
        }

        networks.sortInPlace({ (left, right) -> Bool in
            if left.nwid < right.nwid {
                return true
            }

            return false
        })

        objc_sync_enter(allNetworks)
        allNetworks = networks
        objc_sync_exit(allNetworks)

        saveNetworks()

        let nc = NSNotificationCenter.defaultCenter()

        nc.postNotificationName(networkUpdateKey, object: nil, userInfo: ["networks": networks])
    }

    private func findNetworkWithID(nwid: UInt64) -> Int {
        for (index, element) in allNetworks.enumerate() {

            if element.nwid == nwid {
                return index
            }
        }

        return NSNotFound
    }

    private func findSavedNetworkWithID(nwid: UInt64) -> Int {
        for (index, element) in savedNetworks.enumerate() {

            if element.nwid == nwid {
                return index
            }
        }

        return NSNotFound
    }

    private func saveNetworks() {
        let file = dataFile()

        objc_sync_enter(allNetworks)
        NSKeyedArchiver.archiveRootObject(self.allNetworks, toFile: file)
        objc_sync_exit(allNetworks)
    }
}
